aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -55,7 +55,7 @@ namespace meta {
       mat1.sizes()[0], "x", mat1.sizes()[1], " and ", mat2.sizes()[0], "x", mat2.sizes()[1], ")"); \
  \
   auto names = at::namedinference::propagate_names_for_addmm(mat1, mat2, self); \
-  set_output_raw_strided(0, {mat1.sizes()[0], mat2.sizes()[1]}, {}, self.options(), names);
```

Comment:
should we also check that mat1, mat2 and self dtypes are the same? The actual implementation will error out on the different types.

Reply:
ah, I see. Okay, I can add that check then

Reply:
oh wait, I can add a check to see if the dtypes of `self` and `mat2` match as they are the `bias` and `weight` respectively. However, checking if the `input` dtype matches the parameters will fail in the case that I originally made the issue for

Reply:
@ngimel is this fine?

Reply:
Empirically on CPU with bfloat16 and float32 they have to all match. Are you saying that the CUDA behavior is different?

```
>>> def test(c):
...   try:
...     c()
...     return True
...   except:
...     return False
... 
>>> for i,j,k in itertools.product(*[[torch.float32,torch.bfloat16]]*3): print(i,j,k,test(lambda: torch.addmm(torch.randn(1,1,dtype=i), torch.randn(1,1,dtype=j), torch.randn(1,1,dtype=k))))
... 
torch.float32 torch.float32 torch.float32 True
torch.float32 torch.float32 torch.bfloat16 False
torch.float32 torch.bfloat16 torch.float32 False
torch.float32 torch.bfloat16 torch.bfloat16 False
torch.bfloat16 torch.float32 torch.float32 False
torch.bfloat16 torch.float32 torch.bfloat16 False
torch.bfloat16 torch.bfloat16 torch.float32 False
torch.bfloat16 torch.bfloat16 torch.bfloat16 True
```

Reply:
Cuda addmm behavior is the same, all the dtypes have to match.

Reply:
The case I'm referring to isn't in any execution mode. Its for custom mixed precision on custom vendor hardware that is only _traced_ via lazy tensors. Hence, the case
```
torch.float32 torch.float16 torch.float32
```
should be allowed here and should output type `torch.float16` to match the input. The actual casting is done further down our stack, not at the top level.

Reply:
Ah, then I think I would probably claim your custom AMP is wrong. The two ways I can think of solving this: (1) explicitly generate the casts and fuse then away in your backend compiler, (2) produce an alternative matmul that is type promoting and have the backend target that

Reply:
okay, we've decided to do with solution (1). In which case, the changes in this PR aren't quite necessary anymore. Do we still want to merge this in (including the added dtype checks)?

Reply:
Dtype checks would be appreciated!